File: pyreversi4/player.py
from pyreversi4.ai import *
from pyreversi4.evaluator import *

from abc import *

class Player(object):
    __metaclass__ = ABCMeta

    def __init__(self):
        pass

    @abstractmethod
    def act(self, board):
        raise NotImplementError()

class PlayerRandom(Player):
    def __init__(self, color, name='Random'):
        self.color = color
        self.name = name

    def act(self, board):
        ai = RandomAI()
        ai.move(board)

class PlayerNegaMax(Player):
    def __init__(self, color, name='NegaMax', max_depth=15):
        self.color = color
        self.name = name
        self.max_depth = max_depth

    def act(self, board):
        evaluator = PerfectEvaluator()
        ai = NegaMaxAI(evaluator, self.max_depth)
        ai.move(board)

class PlayerMonteCarlo(Player):
    def __init__(self, color, name='MonteCarlo', try_num=1000):
        self.color = color
        self.name = name
        self.try_num = try_num

    def act(self, board):
        ai = MonteCarloAI(self.try_num)
        ai.move(board)

class PlayerMonteCarloTree(Player):
    def __init__(self, color, name='MonteCarloTree', try_num=1000):
        self.color = color
        self.name = name
        self.try_num = try_num

    def act(self, board):
        ai = MonteCarloTreeAI(self.try_num)
        ai.move(board)

File: cppreversi/Player.cpp
#include "Player.h"
#include "AI.h"

#include <complex>

Player::Player(const Color color, const std::string name) {
    this->color = color;
    this->name = name;
}

PlayerRandom::PlayerRandom(const Color color, const std::string name) : Player(color, name) {

}

void PlayerRandom::act(Board& board) {
    RandomAI ai;
    std::uint64_t hand = ai.act(board);
    board.move(hand);
}

PlayerNegaMax::PlayerNegaMax(const Color color, const std::string name) : Player(color, name) {

}

void PlayerNegaMax::act(Board& board) {
    NegaMaxAI ai;
    std::uint64_t hand = ai.act(board);
    board.move(hand);
}

PlayerNegaScout::PlayerNegaScout(const Color color, const std::string name) : Player(color, name) {

}

void PlayerNegaScout::act(Board& board) {
    NegaMaxAI ai;
    std::uint64_t hand = ai.act(board);
    board.move(hand);
}

PlayerMonteCarlo::PlayerMonteCarlo(const Color color, const std::string name) : Player(color, name) {
    this->trials = 3000;
}

PlayerMonteCarlo::PlayerMonteCarlo(const Color color, const std::string name, const int trials) : Player(color, name) {
    this->trials = trials;
}

void PlayerMonteCarlo::act(Board& board) {
    MonteCarloAI ai(this->trials);
    std::uint64_t hand = ai.act(board);
    board.move(hand);
}

PlayerMonteCarloTree::PlayerMonteCarloTree(const Color color, const std::string name) : Player(color, name) {
    this->trials = 3000;
    this->exploration_param = std::sqrt(2);
    this->playout_threshold = 10;
}

PlayerMonteCarloTree::PlayerMonteCarloTree(const Color color, const std::string name,
    const int trials, const double exploration_param, const int playout_threshold) : Player(color, name) {
    this->trials = trials;
    this->exploration_param = exploration_param;
    this->playout_threshold = playout_threshold;
}

void PlayerMonteCarloTree::act(Board& board) {
    MonteCarloTreeAI ai(this->trials, this->exploration_param, this->playout_threshold);
    std::uint64_t hand = ai.act(board);
    board.move(hand);
}
